pontoon/base/middleware.py

```diff
@@ -34,3 +37,30 @@ def process_request(self, request):
             return HttpResponseForbidden("<h1>Forbidden</h1>")
 
         return None
+
+
+class EmailConsentMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        response = self.get_response(request)
+
+        if not settings.EMAIL_CONSENT_ENABLED:
+            return response
+
+        if not request.user.is_authenticated:
+            return response
+
+        if request.user.profile.email_consent_dismissed_at is not None:
+            return response
+
+        if is_ajax(request):
+            return response
+
+        email_consent_url = "pontoon.messaging.email_consent"
+        if request.path == reverse(email_consent_url):
+            return response
+
+        request.session["next_path"] = request.get_full_path()
+        return redirect(email_consent_url)
```

Comment:
Would it be possible to retain the previous path in a query parameter rather than a session variable? It would make it clearer to a user that they'll get to where they wanted to go.

Reply:
Personally I'm not a big fan of that, I worry the path will get lost if somebody edits the URL, and that it'll get ugly if the previous path contains a query string.